### DISCUSSION TITLE
Best way to message a ring from outside of any ring.

### DISCUSSION THREAD

**Discussion by RedBeard0531** [axboe/liburing#1012] [Q&A] [answered]
I know that there is `IORING_OP_MSG_RING` for sending a message from one ring to another, but is there a similar API for messaging a ring from a context that doesn't have its own ring? I was thinking something like calling `write(ringfd, someSizeBuffer)`, or possibly a dedicated syscall like `io_uring_message(ringfd,  u64 user_data, int res)`. I can build something similar out of an user-space in-memory queue + either an eventfd or (as of 6.7) a futex. But it seemed like something built-in to behave like ring-to-ring messaging would be both simpler and more efficient.

**Comment by axboe** [comment, marked as answer]
This could potentially be added as a write operation indeed, but I don't think we want that extra complexity. Right now an io_uring fd does not support read/write, and adding the ability to do so would require more checking than just simply adding the ability to write to the fd. Hence I'm reluctant to do that, when the app could just setup a ring and use that for messaging.

**Comment by DylanZA** [comment]
I think your solution is about as efficient as you will get. or if you don't care about efficiency as much you could just use a pipe as your in-memory queue and read from that in the io_uring
